### PR TITLE
Add Order

### PR DESCRIPTION
**Pull request by Chorniypistolet**

### CODE REVIEW

File: src/main/java/mate/academy/spring/boot/controller/OrderItemController.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import java.util.List;
+
+@RestController
+@RequestMapping("/orders")
+@RequiredArgsConstructor
+public class OrderItemController {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/OrderItemController.java]: replace all endpoints to OrderController

File: src/main/java/mate/academy/spring/boot/dto/order/OrderRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.spring.boot.dto.order;
+
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class OrderRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/order/OrderRequestDto.java]:
```suggestion
    @NotBlank
```

File: src/main/java/mate/academy/spring/boot/dto/status/StatusRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.spring.boot.dto.status;
+
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class StatusRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/status/StatusRequestDto.java]:
```suggestion
    @NotBlank
```

File: src/main/java/mate/academy/spring/boot/mapper/OrderItemMapper.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import mate.academy.spring.boot.config.MapperConfig;
+import mate.academy.spring.boot.dto.orderItemDto.OrderItemDto;
+import mate.academy.spring.boot.model.OrderItem;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface OrderItemMapper {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/mapper/OrderItemMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/mapper/OrderMapper.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import mate.academy.spring.boot.model.ShoppingCart;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import java.math.BigDecimal;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Mapper(config = MapperConfig.class, uses = OrderItemMapper.class)
+public interface OrderMapper {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/mapper/OrderMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/model/OrderItem.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import jakarta.persistence.Column;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import java.math.BigDecimal;
+
+@Entity
+@Getter
+@Setter
+@ToString
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/OrderItem.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/model/Status.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.spring.boot.model;
+
+public enum Status {
+    PENDING,
+    COMPLETED,
+    CANCELLED,
+    DELIVERED;
+
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/Status.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/model/User.java [hunk 8]
```diff
@@ -49,6 +52,8 @@ public class User implements UserDetails {
             inverseJoinColumns = @JoinColumn(name = "role_id")
     )
     private Set<Role> roles = new HashSet<>();
+    @OneToMany(mappedBy = "user", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
+    private Set<Order> orders;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/User.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/repository/order/OrderRepository.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+import mate.academy.spring.boot.model.Order;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+import java.util.Optional;
+
+public interface OrderRepository extends JpaRepository<Order, Long> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/repository/order/OrderRepository.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/OrderService.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+import mate.academy.spring.boot.dto.order.OrderDto;
+import mate.academy.spring.boot.dto.order.OrderRequestDto;
+import mate.academy.spring.boot.dto.status.StatusRequestDto;
+import mate.academy.spring.boot.model.User;
+
+import java.util.List;
+
+public interface OrderService {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/OrderService.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/impl/OrderItemServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import mate.academy.spring.boot.mapper.OrderItemMapper;
+import mate.academy.spring.boot.model.OrderItem;
+import mate.academy.spring.boot.repository.orderItem.OrderItemRepository;
+import mate.academy.spring.boot.service.OrderItemService;
+import org.springframework.stereotype.Service;
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class OrderItemServiceImpl implements OrderItemService {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/OrderItemServiceImpl.java]: replace logic to OrderService

File: src/main/java/mate/academy/spring/boot/service/impl/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    private final OrderRepository orderRepository;
+    private final OrderMapper orderMapper;
+    private final ShoppingCartRepository shoppingCartRepository;
+
+    @Override
+    public List<OrderDto> getAllOrdersByUserId(Long id) {
+        return orderRepository.findAll().stream()
+                .filter(order -> order.getUser().getId().equals(id))
+                .map(orderMapper :: toDto)
+                .toList();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/OrderServiceImpl.java]:
```suggestion
        return orderMapper.toOrderDtoList(orderRepository.findAllByUserId(userId, pageable));
```

File: src/main/java/mate/academy/spring/boot/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final OrderMapper orderMapper;
+    private final ShoppingCartRepository shoppingCartRepository;
+
+    @Override
+    public List<OrderDto> getAllOrdersByUserId(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/OrderServiceImpl.java]:
```suggestion
    public List<OrderDto> getAllOrdersByUserId(Long id, Pageable pageable) {
```

File: src/main/java/mate/academy/spring/boot/service/impl/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    public OrderDto updateStatus(Long id, StatusRequestDto statusRequestDto) {
+        Order order = orderRepository.findOrderWithItems(id)
+                .orElseThrow(() -> new EntityNotFoundException("No such a order with id: " + id));
+        String statusString = statusRequestDto.getStatus();
+        Status newStatus = Arrays.stream(Status.values())
+                .filter(status -> status.name().equalsIgnoreCase(statusString))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("Invalid status: " + statusString));
+        order.setStatus(newStatus);
+        orderRepository.save(order);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/OrderServiceImpl.java]:
```suggestion
        Order order = orderRepository.findById(orderId)
                .map(o -> {
                    o.setStatus(Status.valueOf(orderDto.status()));
                    return o;
                })
                .orElseThrow(() -> new EntityNotFoundException(
                        String.format("Order with id: %d not found", orderId)
                ));
                return orderMapper.toUpdateDto(orderRepository.save(order));
```

File: src/main/java/mate/academy/spring/boot/service/impl/OrderServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public OrderDto createOrder(Long userId, OrderRequestDto orderRequestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found for user " + userId));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/OrderServiceImpl.java:60]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found for user " + userId));
                if (shoppingCart.getCartItems().isEmpty()) { 
                    throw new OrderProcessingException
```

File: src/main/resources/db/changelog/changes/14-create-orders-table.yaml [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+                    nullable: false
+
+        - addForeignKeyConstraint:
+            baseTableName: orders
+            baseColumnNames: user_id
+            constraintName: fk_order_user
+            referencedTableName: users
+            referencedColumnNames: id
+
+        - createIndex:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/14-create-orders-table.yaml]: remove creating index

File: src/main/resources/db/changelog/changes/16-update-orders-table.yaml [hunk 17]
```diff
@@ -0,0 +1,9 @@
+databaseChangeLog:
+  - changeSet:
+      id: 16
+      author: ChorniyPistolet
+      changes:
+        - modifyDataType:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/16-update-orders-table.yaml]: remove this changeSet and id: 17 also
modify datatype columnName: status in existing changesets

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**